sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataGrpcMultiplexer.java

```diff
@@ -82,16 +73,34 @@ public String toString() {
     return outboundObserver;
   }
 
-  public CompletableFuture<Consumer<BeamFnApi.Elements.Data>> futureForKey(
-      LogicalEndpoint key) {
-    return consumers.computeIfAbsent(key, (LogicalEndpoint unused) -> new CompletableFuture<>());
+  private SettableFuture<DataBytesReceiver> receiverFuture(LogicalEndpoint endpoint) {
+    SettableFuture<DataBytesReceiver> future = consumers.get(endpoint);
+    if (future == null) {
+      future = SettableFuture.create();
+      SettableFuture<DataBytesReceiver> present =
+          consumers.putIfAbsent(endpoint, future);
```

Comment:
computeIfAbsent is so much cleaner :(

Reply:
Yeah, I'm also sad about it

Reply:
Now that JDK 8 is back in, we should be able to use computeIfAbsent